clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -686,6 +694,56 @@ private void producePutOrDeleteToKafka(
           kafkaClusterId,
           beforeProcessingRecordTimestampNs);
     }
+
+    /**
+     * Here we will only delete deprecated chunks if the processed message is UPDATE message and chunking is enabled,
+     * as we do not want to do it for every PUT/DELETE message. It might still leave some lingering chunks though when
+     * old chunked value got replaced either through PUT or DELETE operation.
+     */
+    if (msgType.equals(MessageType.UPDATE) && isChunked) {
+      Delete delete = new Delete();
+      delete.schemaId = AvroProtocolDefinition.CHUNK.getCurrentProtocolVersion();
+      delete.replicationMetadataVersionId = rmdProtocolVersionID;
+      delete.replicationMetadataPayload = WriterChunkingHelper.EMPTY_BYTE_BUFFER;
+      DeleteMetadata deleteMetadata = new DeleteMetadata(
+          AvroProtocolDefinition.CHUNK.getCurrentProtocolVersion(),
+          rmdProtocolVersionID,
+          WriterChunkingHelper.EMPTY_BYTE_BUFFER);
+      if (oldValueManifest != null) {
+        for (int i = 0; i < oldValueManifest.keysWithChunkIdSuffix.size(); i++) {
+          LeaderProducedRecordContext valueChunkDeleteRecordContext =
+              LeaderProducedRecordContext.newChunkDeleteRecord(updatedKeyBytes, delete);
+          byte[] chunkKeyBytes = oldValueManifest.keysWithChunkIdSuffix.get(i).array();
+          produceToLocalKafka(
+              consumerRecord,
+              partitionConsumptionState,
+              valueChunkDeleteRecordContext,
+              (callback, leaderMetadataWrapper) -> veniceWriter.get()
+                  .deleteDeprecatedChunk(chunkKeyBytes, subPartition, callback, leaderMetadataWrapper, deleteMetadata),
+              subPartition,
+              kafkaUrl,
+              kafkaClusterId,
+              beforeProcessingRecordTimestampNs);
+        }
+      }
+      if (oldRmdManifest != null) {
+        for (int i = 0; i < oldRmdManifest.keysWithChunkIdSuffix.size(); i++) {
+          LeaderProducedRecordContext rmdChunkDeleteRecordContext =
+              LeaderProducedRecordContext.newChunkDeleteRecord(updatedKeyBytes, delete);
+          byte[] chunkKeyBytes = oldRmdManifest.keysWithChunkIdSuffix.get(i).array();
+          produceToLocalKafka(
+              consumerRecord,
+              partitionConsumptionState,
+              rmdChunkDeleteRecordContext,
+              (callback, leaderMetadataWrapper) -> veniceWriter.get()
+                  .deleteDeprecatedChunk(chunkKeyBytes, subPartition, callback, leaderMetadataWrapper, deleteMetadata),
+              subPartition,
+              kafkaUrl,
+              kafkaClusterId,
+              beforeProcessingRecordTimestampNs);
+        }
```

Comment:
I am wondering whether this logic belongs to `VeniceWriter` and `LeaderProducerCallback` or not.
For example, we can add a new param called `byte[] previousChunks` in `VeniceWriter#put` method and VeniceWriter will send these deletion messages to Kafka internally, and in the callback, StoreIngestionTask will produce the chunk deletion messages into buffer drainer.
IMO, this structure makes more sense because of modular structure and reusable, so that we don't need to repeat the same logic in `LeaderFollowerStoreIngestionTask`.
What do you think?

Reply:
Sure, let me see if I can do that. 

Previously I was struggling to produce multiple chunk delete within a single VW call as it somehow has issue with Leader produce context and such...but let me try to re-org the code...

Reply:
Hmm, in my latest changes, I actually move the for loop logic into LFSIT#produceToLocalKafka. 
But I think what you said makes more sense...let me see how I can move it into VW.
(And I will have to change VW a bit, as it does not have information of rmdProtocolVersionId, which makes it pretty ugly to pass along some static information all the value in the call chain).

Maybe I will change PUT API to carry old manifests, and change ChunkAwareCallback#setChunkingInfo to add two args so that in the LPC it can iterate over them and produce them to Leader drainer as well.